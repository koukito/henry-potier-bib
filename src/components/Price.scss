.price__before-discount {
  text-decoration: line-through;
  margin-right: 1rem;
  margin-left: 1rem;
  font-weight: bold;
}
.price-string {
  font-weight: bold;
  display: flex;
}
