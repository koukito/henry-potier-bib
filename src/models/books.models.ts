export interface Book {
  isbn: string;
  title: string;
  price: number;
  cover: string;
  synopsis: Array<string>;
  quantity?: number;
}

export interface Offer {
  type: OfferType;
  value: number;
  sliceValue?: number;
}

export type OfferType = 'percentage' | 'minus' | 'slice';
