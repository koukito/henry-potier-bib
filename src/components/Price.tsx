import React from 'react';
import calculateDiscountPrice from '../helpers/calculateDiscountPrice';
import { Book, Offer } from '../models/books.models';
import { Row, Col } from 'react-bootstrap';
import './Price.scss';
export interface PriceProps {
  offers: Array<Offer>;
  cartItems: Array<Book>;
}
const Price = ({ offers, cartItems }: PriceProps) => {
  const totalPriceBeforeDiscount = cartItems
    .map((book) => book.price)
    .reduce((prev, current) => {
      return prev + current;
    }, 0);
  const pricesAfterDiscount = offers
    .map((offer: Offer) => ({
      type: offer.type,
      priceAfterDiscount: calculateDiscountPrice(
        offer,
        totalPriceBeforeDiscount
      )
    }))
    .sort((a, b) => (a.priceAfterDiscount > b.priceAfterDiscount ? 1 : -1));
  const priceAfterDiscount = pricesAfterDiscount[0].priceAfterDiscount;

  return (
    <Row>
      <Col sm={10}></Col>
      <Col sm={2}>
        <span className="price-string">
          Total price
          <span className="price__before-discount">
            {totalPriceBeforeDiscount}€
          </span>
          {priceAfterDiscount}€
        </span>
      </Col>
    </Row>
  );
};
export default Price;
