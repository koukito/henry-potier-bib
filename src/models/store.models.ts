import { Book } from './books.models';

export type Action =
  | { type: 'ADD_BOOK_CART'; value: Book }
  | { type: 'REMOVE_BOOK_CART'; value: string }
  | { type: 'SET_SEARCH_TERM'; value: string };

export type Dispatch = (action: Action) => void;

export interface BooksStoreState {
  cartItems: Array<Book>;
  searchTerm: string;
}
