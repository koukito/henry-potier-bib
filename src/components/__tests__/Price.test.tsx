import React from 'react';
import renderer from 'react-test-renderer';
import { Offer } from '../../models/books.models';
import Price from '../Price';

const mockBookData1 = {
  isbn: 'mock1',
  title: 'MockTitlte',
  synopsis: [],
  price: 35,
  cover: '',
  quantity: 1
};
const mockBookData2 = {
  isbn: 'mock2',
  title: 'MockTitlte',
  synopsis: [],
  price: 30,
  cover: '',
  quantity: 1
};
const mockCartItems = [mockBookData1, mockBookData2];
const mockOffersData: Array<Offer> = [
  { type: 'percentage', value: 5 },
  { type: 'minus', value: 15 },
  { type: 'slice', sliceValue: 100, value: 12 }
];

describe('Price', () => {
  it('renders correctly the right prices', () => {
    const tree = renderer
      .create(<Price offers={mockOffersData} cartItems={mockCartItems} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
