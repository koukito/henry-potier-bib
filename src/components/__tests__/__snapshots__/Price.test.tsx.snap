// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price renders correctly the right prices 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-10"
  />
  <div
    className="col-sm-2"
  >
    <span
      className="price-string"
    >
      Total price
      <span
        className="price__before-discount"
      >
        65
        €
      </span>
      50
      €
    </span>
  </div>
</div>
`;
